### ISSUE TITLE
Use without installer?

### ISSUE THREAD

**Issue by Sledro** [vslavik/winsparkle#142]
Is there a way to use this to simply replace a portable exe without an installer?

**Comment by vslavik** [comment]
Not currently. PRs welcome.

**Comment by solusipse** [comment]
@Sledro We use Winsparkle without an installer - you could use [Windows' resource system](https://msdn.microsoft.com/en-us/library/cc194804.aspx) to store a zip archive within an executable. After downloading this executable, Winsparkle will run it in application's directory. You could simply unpack your new files there. I can provide some examples if you want. I could also provide a PR, but that would introduce new dependency (something to handle zips) - I don't know if that's allowed by Winsparkle's policy. @vslavik what are your thoughts about that?

**Comment by vslavik** [comment]
> We use Winsparkle without an installer - you could use Windows' resource system to store a zip archive within an executable.

In other words, what's the term for such an executable… ah, an *installer.*

> Winsparkle will run it in application's directory

It will do no such thing. Downloads are put into a random temporary directory.

>  I don't know if that's allowed by Winsparkle's policy.

I would naively think that just looking at the sources and seeing the `3rdparty` directory is a pretty clear indication...

**Comment by solusipse** [comment]
>It will do no such thing. Downloads are put into a random temporary directory.

You're actually wrong Vaclav! How does it matter where you put your downloads if the working directory is different (application's directory in this case, maybe you're not aware, Winsparkle works in this way)?

**Comment by vslavik** [comment]
> You're actually wrong Vaclav! 

Calm down. First, you can't expect people to understand that when you say "run in X", you don't *actually mean* "run in X" (i.e. place in X and run), but "run somewhere else entirely while pointing CWD to X". Second, unpacking into CWD w/o no checks would be a rookie mistake. Third (and the reason for the previous point), current CWD value may be literally *anything* — depending on what the user code, which WinSparkle has no control over, does and even depending *on the way the application was launched.*

So no, I'm not wrong and I'm well aware of WinSparkle's internals, thans for asking.

**Comment by solusipse** [comment]
Now you just added some extra assumptions and provided examples of some limitations that could be easily overcome. But never mind, I see that you don't like this idea and I'm not going to fight you about it. Take care and have a good day!

**Comment by vslavik** [comment]
@solusipse I don't know whether you're twisting my words into their exact opposite or are just that bad with reading comprehension, but I said no such thing. I do take issue with you spreading dangerous misinformation and advising people to do insecure things here, so I had to set the record straight. That has exactly nothing to do with whether unpacking ZIPs is welcome or not (there's a reason this issue is still open, you know).

**Comment by solusipse** [comment]
@vslavik I understand that. You had some good points and I can often agree with you! I just don't see why you have to act like you're the new Linus Torvalds. I took a brief look at your other responses and I see that you usually react with impoliteness to people who want to contribute or report issues to you. You don't have to be kind, of course. Just be tolerable Vaclav! Take care!

**Comment by vslavik** [comment]
@solusipse Says the person who says complete nonsense here and that rudely accuses the developer of something he uses that he is clueless. The same person that files a vague, badly described issue, and can't even be bothered to respond on it. That's rich.

Sorry that I don't have time to carefully tiptoe around entitled people like you in fear of hurting their precious feelings. I have actual work to do — the one you don't mind using.

**Comment by vslavik** [comment]
[Deleted a series of factually wrong and increasingly off-topic comments.]

**Comment by bvgastel** [comment]
I am interested in this feature. Any pointers how to implement this feature? I can see of I can help (not a Windows developer though).

**Comment by codygray** [comment]
It would be relatively straightforward to do this on Windows. The key is that, while you cannot *delete* a running executable, you can easily **rename** it.

So, once an update is initiated, the workflow would be:

 - Have WinSparkle rename the currently-running EXE by appending ".bak" to its name (or whatever).
 - Have WinSparkle download the new (updated) EXE, storing it in the same folder as the currently-running EXE with the original name of the currently-running EXE (i.e., without the ".bak" suffix).

The currently-running application won't be updated *live*---the user will need to close the application and re-open it. (WinSparkle could theoretically force the application to close after it completed downloading the new EXE, but this seems needlessly invasive.) But when the user does go to re-open the application, because the updated EXE has been put in the place of the original EXE, the process will be seamless to them.

For completeness, code should be added to WinSparkle that checks for any ".bak" files in the application folder and removes them. These are vestiges of the previous update process, and should no longer be required. (Bonus: A copy of the old application binary is kept until the updated application binary successfully starts up.)

cc @bvgastel , in case you're still interested in implementing this feature. Armed with this information, you should not need any particular Windows-programming expertise. All you need to know how to do is how to rename and delete files.

**Comment by vslavik** [comment]
> The key is that, while you cannot delete a running executable, you can easily rename it.

Yep, that's crucial. Assuming the same applies to DLLs and `WinSparkle.dll` can be renamed while loaded, then in-place patching is relatively straightforward too.
